smack-extensions/src/main/java/org/jivesoftware/smackx/nick/NickManager.java

```diff
@@ -0,0 +1,127 @@
+import org.jivesoftware.smack.XMPPConnection;
+import org.jivesoftware.smack.filter.StanzaFilter;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.packet.Stanza;
+import org.jivesoftware.smackx.nick.filter.NickFilter;
+import org.jivesoftware.smackx.nick.packet.Nick;
+
+import org.jxmpp.jid.EntityBareJid;
+
+
+/**
+ * Implementation of XEP-0172.
+ *
+ * @author Miguel Hincapie 2018
+ * @see <a href="http://xmpp.org/extensions/xep-0172.html">XEP-0172: User Nickname</a>
+ */
+public final class NickManager extends Manager {
+
+    private static final Map<XMPPConnection, NickManager> INSTANCES = new WeakHashMap<>();
+
+    private static final StanzaFilter INCOMING_MESSAGE_FILTER = NickFilter.INSTANCE;
+
+    private final Set<NickListener> nickListeners = new HashSet<>();
+
+    private final AsyncButOrdered<Message> asyncButOrdered = new AsyncButOrdered<>();
+
+    private NickManager(XMPPConnection connection) {
+        super(connection);
+
+        connection.addSyncStanzaListener(new StanzaListener() {
+            @Override
+            public void processStanza(Stanza packet)
+                    throws
+                    SmackException.NotConnectedException,
+                    InterruptedException,
+                    SmackException.NotLoggedInException {
+                final Message message = (Message) packet;
+
+                asyncButOrdered.performAsyncButOrdered(message, new Runnable() {
+                    @Override
+                    public void run() {
+                        for (NickListener listener : nickListeners) {
+                            listener.newNickMessage(message);
+                        }
+                    }
+                });
+
+            }
+        }, INCOMING_MESSAGE_FILTER);
+    }
+
+    public static synchronized NickManager getInstanceFor(XMPPConnection connection) {
+        NickManager nickManager = INSTANCES.get(connection);
+        if (nickManager == null) {
+            nickManager = new NickManager(connection);
+            INSTANCES.put(connection, nickManager);
+        }
+        return nickManager;
+    }
+
+    public synchronized boolean addNickMessageListener(NickListener listener) {
+        return nickListeners.add(listener);
+    }
+
+    public synchronized boolean removeNickMessageListener(NickListener listener) {
+        return nickListeners.remove(listener);
+    }
+
+    public void sendNickMessage(EntityBareJid to, String nickname) throws
+            SmackException.NotLoggedInException,
+            InterruptedException,
+            SmackException.NotConnectedException {
+        sendStanza(createNickMessage(to, nickname));
+    }
+
+    /**
+     * Create a Smack's message stanza to update the user's nickName.
+     *
+     * @param to       the receiver.
+     * @param nickName the new nickName.
+     * @return instance of Message stanza.
+     */
+    private Message createNickMessage(EntityBareJid to, String nickName) {
```

Comment:
There is no reason this method should not be declared static.

Reply:
done, static added